test/external-modules/apm-integration/src/javaRestTest/java/org/elasticsearch/test/apmintegration/MetricsApmIT.java

```diff
@@ -91,21 +88,35 @@ public void testApmIntegration() throws Exception {
                 var samples = (Map<String, Object>) metricset.get("samples");
 
                 samples.forEach((key, value) -> {
-                    var assertion = sampleAssertions.get(key);// sample name
-                    if (assertion != null) {
-                        logger.info("Matched {}", key);
+                    var valueAssertion = valueAssertions.get(key);// sample name
+                    if (valueAssertion != null) {
+                        logger.info("Matched {}:{}", key, value);
                         var sampleObject = (Map<String, Object>) value;
-                        if (assertion.test(sampleObject)) {// sample object
+                        if (valueAssertion.test(sampleObject)) {// sample object
+                            logger.info("{} assertion PASSED", key);
+                            valueAssertions.remove(key);
+                        } else {
+                            logger.error("{} assertion FAILED", key);
+                        }
+                    }
+                    var histogramAssertion = histogramAssertions.get(key);
+                    if (histogramAssertion != null) {
+                        logger.info("Matched {}:{}", key, value);
+                        var samplesObject = (Map<String, Object>) value;
+                        var counts = ((Collection<Integer>) samplesObject.get("counts")).stream().mapToInt(Integer::intValue).sum();
+                        var remaining = histogramAssertion - counts;
+                        if (remaining == 0) {
                             logger.info("{} assertion PASSED", key);
-                            sampleAssertions.remove(key);
+                            histogramAssertions.remove(key);
                         } else {
-                            logger.error("{} assertion FAILED: {}", key, sampleObject.get("value"));
+                            logger.info("{} assertion PENDING: {} remaining", key, remaining);
+                            histogramAssertions.put(key, remaining);
```

Comment:
I think this test will fail to notice if there are too many samples with the given `key`, because as soon as the count hits zero, the entry is removed, and thereafter, we do no more checks. The existing logic also had this bug.

Also, style-wise, this could use `map.computeIfPresent` if you find that cleaner at all:

```
histogramAssertions.computeIfPresent(key, (k, old) -> {
    int counts = WHATEVER;
    var remaining = old - counts;
    if (remaining == 0) {
        return null;
    } else {
        return remaining;
    }
});
```